test/integration/etcd_storage_path_test.go

```diff
@@ -0,0 +1,599 @@
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/util/flowcontrol"
+
+	"github.com/coreos/etcd/clientv3"
+
+	appserver "github.com/kubernetes-incubator/service-catalog/cmd/apiserver/app/server"
+	"github.com/kubernetes-incubator/service-catalog/pkg/api"
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
+	registryserver "github.com/kubernetes-incubator/service-catalog/pkg/registry/servicecatalog/server"
+)
+
+// Etcd data for all persisted objects.
+var etcdStorageData = map[schema.GroupVersionResource]struct {
+	stub             string                   // Valid JSON stub to use during create
+	prerequisites    []prerequisite           // Optional, ordered list of JSON objects to create before stub
+	expectedEtcdPath string                   // Expected location of object in etcd, do not use any variables, constants, etc to derive this value - always supply the full raw string
+	expectedGVK      *schema.GroupVersionKind // The GVK that we expect this object to be stored as - leave this nil to use the default
+}{
+	gvr("servicecatalog.k8s.io", "v1beta1", "clusterservicebrokers"): {
+		stub:             `{"metadata": {"name": "broker1"}, "spec": {"url": "https://broker1.com"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/clusterservicebrokers/broker1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "clusterserviceclasses"): {
+		stub:             `{"metadata": {"name": "class1"}, "spec": {"clusterServiceBrokerName": "broker1", "externalName": "class-name1", "externalID": "class1", "description": "desc"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/clusterserviceclasses/class1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "clusterserviceplans"): {
+		stub:             `{"metadata": {"name": "plan1"}, "spec": {"clusterServiceBrokerName": "broker1", "externalName": "plan-name1", "externalID": "plan1", "description": "desc", "clusterServiceClassRef": {"Name": "class1"}}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/clusterserviceplans/plan1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "serviceinstances"): {
+		stub:             `{"metadata": {"namespace": "etcdstoragepathtestnamespace", "name": "instance1"}, "spec": {"clusterServiceClassExternalName": "class1", "clusterServicePlanExternalName": "plan1", "externalID": "instance1"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/serviceinstances/etcdstoragepathtestnamespace/instance1",
+	},
+	gvr("servicecatalog.k8s.io", "v1beta1", "servicebindings"): {
+		stub:             `{"metadata": {"namespace": "etcdstoragepathtestnamespace", "name": "binding1"}, "spec": {"instanceRef": {"name": "instance1"}, "externalID": "binding1"}}`,
+		expectedEtcdPath: "/servicecatalog.k8s.io/servicebindings/etcdstoragepathtestnamespace/binding1",
+	},
+}
+
+// Be very careful when whitelisting an object as ephemeral.
+// Doing so removes the safety we gain from this test by skipping that object.
+var ephemeralWhiteList = createEphemeralWhiteList()
```

Comment:
What is ephemeral? Can we skip having any of this code in lieu of linking to upstream? Especially having a warning and having none of these objects.

Reply:
An ephemeral resource is one that is not stored in etcd.